### PR TITLE
memory: Hybrid context shift

### PR DESCRIPTION
**Pull request by gabe-l-hart**
Closes #16768 

cc @leok7v

## Description

This PR addresses context shift failure caused when a hybrid-recurrent model hits its context limit and attempts to perform context shifting. The main change here is to loosen the restriction in the `llama_memory_recurrent::seq_rm` to only refuse to do a partial erasure if the part being erased includes the final token in the sequence. Since recurrent states are fixed size, any partial erasure that does not include the final token can be considered a no-op.

## Testing

To validate the result, you can use the following which artificially limits the context length to force a context shift:

```sh
# You can use any granite-4.0 model here
./bin/llama-cli -m ggml-org/granite-4.0-h-small-Q8_0-GGUF --jinja -c 100 --context-shift -p "tell me a story"
```

Without this fix, it will fail with `init_batch: failed to prepare attention ubatches`, but with this fix, it will successfully continue generating _and_ produce generated output that is relevant to the previous context.

### CODE REVIEW

File: src/llama-memory-recurrent.cpp [hunk 1]
```diff
@@ -160,8 +161,8 @@ bool llama_memory_recurrent::seq_rm(llama_seq_id seq_id, llama_pos p0, llama_pos
         int32_t & tail_id = cells[seq_id].tail;
         if (tail_id >= 0) {
             const auto & cell = cells[tail_id];
-            // partial intersection is invalid
-            if ((0 < p0 && p0 < cell.pos) || (0 < p1 && p1 <= cell.pos)) {
+            // partial intersection is invalid if it includes the final pos
+            if ((0 < p0 && p1 > cell.pos)) {
```
Comment (compilade) [inline src/llama-memory-recurrent.cpp]: Note that if both `p0` and `p1` are bigger than the final pos, even though technically the state is unaffected, this new condition would still return `false`.

I think `p0` needs to be between 1 and `cell.pos` for the partial intersection case to really happen.

```suggestion
            if ((0 < p0 && p0 <= cell.pos && p1 > cell.pos)) {
```

This could also technically become a sub-case of the tail invalidation check from below, to simplify the logic.

Comment (gabe-l-hart) [reply src/llama-memory-recurrent.cpp]: Ah, good catch!

File: src/llama-memory-recurrent.cpp [hunk 2]
```diff
@@ -160,8 +161,8 @@ bool llama_memory_recurrent::seq_rm(llama_seq_id seq_id, llama_pos p0, llama_pos
         int32_t & tail_id = cells[seq_id].tail;
         if (tail_id >= 0) {
             const auto & cell = cells[tail_id];
-            // partial intersection is invalid
-            if ((0 < p0 && p0 < cell.pos) || (0 < p1 && p1 <= cell.pos)) {
+            // partial intersection is invalid if it includes the final pos
+            if ((0 < p0 && p0 <= cell.pos && p1 > cell.pos)) {
```
Comment (ggerganov) [inline src/llama-memory-recurrent.cpp]:
```suggestion
            if (0 < p0 && p0 <= cell.pos && p1 > cell.pos) {
```

Comment (ggerganov) [inline src/llama-memory-recurrent.cpp]: Why do we check strictly larger than 0 rather than `0 <= p0`?

Comment (gabe-l-hart) [reply src/llama-memory-recurrent.cpp]: If the range includes the first token and extends beyond the last token, it's no longer a partial intersection and can simply clear the whole sequence which is a valid operation here since the answer is just to zero-out everything. The case this is trying to catch is an attempt to go part-way back in the sequence which is invalid since those states are not preserved.

Comment (compilade) [reply src/llama-memory-recurrent.cpp]: > Why do we check strictly larger than 0 rather than `0 <= p0`?

Because when `p0 == 0` (and `p1 > cell.pos`), then the sequence can be removed entirely, since there is no partial intersection.

`p0` larger than 0 means there's at least one remaining token in the context, which would require state rollback.

Comment (gabe-l-hart) [reply src/llama-memory-recurrent.cpp]: (jinx!)

((edit: I hope that translates correctly as "we said the same thing at the same time" to people that don't have little kids))

File: tools/main/main.cpp [hunk 3]
```diff
@@ -354,7 +354,10 @@ int main(int argc, char ** argv) {
         }
 
         // remove any "future" tokens that we might have inherited from the previous session
-        llama_memory_seq_rm(mem, -1, n_matching_session_tokens, -1);
+        if (!llama_memory_seq_rm(mem, -1, n_matching_session_tokens, -1)) {
+            LOG_INF("%s: unable to resuse common prefix\n", __func__);
+            llama_memory_seq_rm(mem, -1, -1, -1);
```
Comment (compilade) [inline tools/main/main.cpp:360]: I'm wondering if this will correctly trigger reevaluation of the context afterwards.

`n_matching_session_tokens` seems to be used *mostly* for cosmetic purposes except when it's equal to `embd_inp.size()`, so maybe it could be set to `0`.

But I think there might be something else necessary here. (For example, should `session_tokens` be cleared?)

(This is mostly relevant when reloading a saved session, and not necessarily during context shifting, though)

**Review by ggerganov — APPROVED** [review]

File: tools/main/main.cpp (hunk 3 above)
Comment (gabe-l-hart) [reply tools/main/main.cpp:360]: Hm, yeah, I'm still a little murky on the logic going on here. I'll look more carefully and see what I can find.

Comment (leok7v) [reply tools/main/main.cpp:360]: Sorry folks,
I am new to merge management of commits.
[b8f9c56923ab34bf24d4fe5476efe9b02736918d](https://github.com/ggml-org/llama.cpp/blob/b8f9c56923ab34bf24d4fe5476efe9b02736918d/src/llama-memory-recurrent.cpp)
Don't know who and when presses merge button for the commit after all checks are passed?

Comment (gabe-l-hart) [reply tools/main/main.cpp:360]: I think we're holding on this until I resolve the question in this thread. I'll try to get to that today!

Comment (gabe-l-hart) [reply tools/main/main.cpp:360]: Ok, I see what you're getting at here with the clause below that recalculates the cached logits for the final token. I think the right thing to do here is to set `n_matching_session_tokens` to `0` since we're essentially saying there are no matching tokens anymore.

File: src/llama-memory-recurrent.cpp (hunk 2 above)
Comment (CISC) [reply src/llama-memory-recurrent.cpp]: `Snap!` is the British equivalent. :)

Edit: (and for some reason `skylder meg en kald cola` (owe me a cold coke) the Norwegian one)

**Comment by gabe-l-hart** [comment]
I've seen this same set of tests failing on a bunch of PRs, so I'm _hoping_ this is just runner instability?

**Comment by CISC** [comment]
> I've seen this same set of tests failing on a bunch of PRs, so I'm _hoping_ this is just runner instability?

Yes. (though I've never seen `SUM` fail before)

**Comment by CISC** [comment]
Is this awaiting final approval, or is it good to merge?

**Review by compilade — APPROVED** [review]

**Review by leok7v — APPROVED** [review]
Looks good to me. Tested. Works.

**Comment by gabe-l-hart** [comment]
@CISC @ggerganov I think with the current approvals, this should be good to go unless you have other outstanding concerns.

**Comment by pwilkin** [comment]
@CISC Don't know if it matters, but I ran into SUM failing for me when doing backwards pass gradient tests for EXPM1 and SOFTPLUS.
